### DISCUSSION TITLE
TTF to OTF conversion

### DISCUSSION THREAD

**Discussion by ftCLI** [fonttools/fonttools#2836] [Q&A] [answered]
I am trying to write a command line TTF to OTF converter as follows:

1. Save the TTF file as TTX (TTF to UFO conversion fails sometimes, but I found that TTX to UFO conversion works)
2. Convert the TTX file to UFO (defcon)
3. Convert the UFO to OTF (ufo2ft)
4. Dump the GDEF, GPOS and GSUB table from the TTF file and merge them into the OTF file. I tried exporting the features with FontFeatures, but importing from the .fea file often fails (tested with IBM Plex).

When saving the OTF file, fontTools prints the following warning:

`GSUB/GPOS Coverage is not sorted by glyph ids.` 

For GSUB table, this code seems to work:

```
if "GSUB" in otf_font:
     print("Reordering GSUB Lookups")
     for lookup in otf_font["GSUB"].table.LookupList.Lookup:
         for sub_table in lookup.SubTable:
             for k, v in sub_table.__dict__.items():
                 if isinstance(fontTools.ttLib.tables.otTables.Coverage, k):
                   for coverage in v:
                       glyphs = coverage.glyphs
                       sorted_glyph_ids = sorted(otf_font.getGlyphIDMany(glyphs))
                       sorted_glyphs = otf_font.getGlyphNameMany(sorted_glyph_ids)
                       coverage.glyphs = sorted_glyphs
```

Is there a faster and more elegant way to process all the `fontTools.ttLib.tables.otTables.Coverage` tables in one pass?

**Comment by justvanrossum** [comment, marked as answer]
I can't answer your question, but I feel you're going about the conversion the wrong way.

The font doesn't need to change much: just convert the glyf outlines to CFF outlines. See `ttFont.getGlyphSet()` for outline extraction and `fontTools.fontBuilder.FontBuilder` for how to build a CFF table with T2 outlines. Make sure to keep the same glyph order, and not a single byte needs changing in GSUB/GPOS and many other tables.

**Comment by ftCLI** [comment]
Thanks, very interesting. For sure wil be faster than ttf -> ttx -> ufo -> otf.

I fear that `addOpenTypeFeatures` will have crash the same way.

**Comment by justvanrossum** [comment]
> I fear that addOpenTypeFeatures will have crash the same way.

The point is, you don't need to decompile/recompile any features if the glyph order stays the same. You can literally use the same GSUB/GPOS binary data, and there is no need to call `addOpenTypeFeatures` in the first place.

**Reply by ftCLI** [reply]
If I just use the same GSUB/GPOS binary data, when FontTools saves the OTF font and compiles the Coverage tables prints the warning

`GSUB/GPOS Coverage is not sorted by glyph ids.`

For what I understand, in OTF fonts the Coverage tables should have glyphs sorted by glyph ids, while in TrueType files is not necessary.

**Reply by justvanrossum** [reply]
> For what I understand, in OTF fonts the Coverage tables should have glyphs sorted by glyph ids, while in TrueType files is not necessary.

This is not true at all.

This either happens because the *original* font already contains broken coverage tables, *or* you've changed the glyph order.

**Comment by ftCLI** [comment]
Definitely I changed the glyphs order. In the original file that warning doesn’t appear. Thanks for your replies, the FontBuilder way will be faster for sure. I’m just trying to understand how to draw the glyphs, but copying the other tables is very easy

**Comment by ftCLI** [comment]
For the record, to suppress the `GSUB/GPOS Coverage is not sorted by glyph ids.` warning:

`destination_font.setGlyphOrder(source_font.getGlyphOrder())`

On the other side, when I try to convert the font using FontBuilder:

```
from fontTools.ttLib.ttFont import TTFont
from fontTools.fontBuilder import FontBuilder, _hheaDefaults, _OS2Defaults
from fontTools.pens.t2CharStringPen import T2CharStringPen
from fontTools.misc.cliTools import makeOutputFileName

ttf_file ="font.ttf"
otf_file = makeOutputFileName(file, extension=".otf")

ttf_font = TTFont("font.ttf")

fb = FontBuilder(unitsPerEm=1000, isTTF=False)

fb.setupGlyphOrder(ttf_font.getGlyphOrder())

fb.setupCharacterMap(ttf_font.getBestCmap())

char_strings = {}
gs = ttf_font.getGlyphSet()
for k, v in gs.items():
    pen = T2CharStringPen(v.width, glyphSet=gs)
    v.draw(pen)
    char_string = pen.getCharString()
    char_strings[k] = char_string

fb.setupCFF(["Test"], charStringsDict=char_strings, fontInfo={}, privateDict={})

fb.setupHorizontalMetrics(ttf_font["hmtx"].metrics)

fb._initTableWithValues("hhea", defaults=_hheaDefaults, values=ttf_font["hhea"].__dict__)

name_strings = {name.nameID: name.toUnicode() for name in ttf_font.name_table.names}

fb.setupNameTable(nameStrings=name_strings)

os2_values = {k: v for k, v in ttf_font["OS/2"].__dict__.items()}
fb._initTableWithValues(tableTag="OS/2", defaults=_OS2Defaults, values=os2_values)

fb.setupPost()

fb.save(otf_file)
```
It fails when saving the output file:

```
Traceback (most recent call last):
  File "D:\PythonProjects\ftCLI-venv\Scripts\ftcli-script.py", line 33, in <module>
    sys.exit(load_entry_point('ftCLI', 'console_scripts', 'ftcli')())
  File "d:\pythonprojects\ftcli\ftCLI\ftCLI.py", line 32, in main
    cli()
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\click\core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\click\core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\click\core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\click\core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\click\core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\click\core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "d:\pythonprojects\ftcli\ftCLI\commands\ftcli_converter.py", line 199, in makeotf
    fb.save(otf_file)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\fontBuilder.py", line 349, in save
    self.font.save(file)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\ttLib\ttFont.py", line 185, in save
    writer_reordersTables = self._save(tmp)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\ttLib\ttFont.py", line 225, in _save
    self._writeTable(tag, writer, done, tableCache)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\ttLib\ttFont.py", line 654, in _writeTable
    self._writeTable(masterTable, writer, done, tableCache)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\ttLib\ttFont.py", line 658, in _writeTable
    tabledata = self.getTableData(tag)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\ttLib\ttFont.py", line 680, in getTableData
    return self.tables[tag].compile(self)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\ttLib\tables\C_F_F_.py", line 19, in compile
    self.cff.compile(f, otFont, isCFF2=False)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\cffLib\__init__.py", line 271, in compile
    writer.toFile(file)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\cffLib\__init__.py", line 522, in toFile
    item.toFile(file)
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\cffLib\__init__.py", line 612, in toFile
    data = tobytes(item, encoding="latin1")
  File "D:\PythonProjects\ftCLI-venv\lib\site-packages\fontTools\misc\textTools.py", line 133, in tobytes
    return bytes(s)
TypeError: 'str' object cannot be interpreted as an integer
```

**Comment by justvanrossum** [comment]
- You don't need to "convert" the whole font, you just need to add a CFF table and delete the glyf table
- FontBuilder takes an optional `font` as argument, which should be a TTFont instance; you can use it to add stuff to an *existing* font
- Do not use internal FontBuilder methods (that start with an underscore): the way you try to use that is not how it works. You cannot take a raw `__dict__` and pass it to some internal method and expect it to do anything reasonable.

**Comment by ftCLI** [comment]
Thanks, works fine even if it produces a bigger CFF table than the reference OTF.

What caused the CFF table compilation error was this line, where I enclosed "Test" into brackets:

`fb.setupCFF(["Test"], charStringsDict=char_strings, fontInfo={}, privateDict={})`

Correct:

`fb.setupCFF("Test", charStringsDict=char_strings, fontInfo={}, privateDict={})`

**Reply by justvanrossum** [reply]
This produces unsubroutinized CFF, so you can likely shrink your file by running a subroutinizer on the built font, such as https://github.com/adobe-type-tools/cffsubr.

**Reply by ftCLI** [reply]
Thanks indeed. It works perfectly and produces better files than method I used in first attempt.

**Comment by ftCLI** [comment]
I'm trying to simplify of outlines of the converted OTF font.

The steps are the following:

- Draw the glyph with pathops.Path() to remove overlaps and fix contours direction
- Get the T2Charstrings to pass to FontBuilder.setupCFF()
- Delete the specific TrueType tables
- With FontBuilder, setup the CFF, maxp and post tables
- Eventually, remove the NULL and CR glyphs
- Subroutinize the font
- Save the font
- Eventually, check the outlines with checkoutlinesufo (it may remove few redundant points)

Here's a comparison of the original TTF (left) file with the converted one (right)

![ttf_vs_t2](https://user-images.githubusercontent.com/83063506/206691234-1bf16748-2095-4891-8fa6-896e5b50e224.PNG)

And here's the converted file (right) compared with the same font converted with makeotf (left)

![afdko_vs_t2](https://user-images.githubusercontent.com/83063506/206691408-bbb70379-54d5-44fe-b74e-4f1d9708d91c.PNG)

I'm aware that the contours can be simplified with FontForge, but I was wondering if there's a pure Python solution to simplify the outlines and get a result similar to the one produced by makeotf.

**Reply by anthrotype** [reply]
I had this piece of code to do qu2qu (quadratic to cubic) conversion but I never got around finishing it, maybe you can find it useful https://github.com/googlefonts/cu2qu/pull/29

**Reply by anthrotype** [reply]
also check this discussion https://github.com/Pomax/bezierjs/issues/151

**Reply by ftCLI** [reply]
Thanks indeed, I'll check. As a temporary workaround, I convert the TTF file to a temp T1 with tx and then the T1 to a temp OTF with makeotf. If tx didn't change the glyph names, it's easy to use the charstrings from the temp OTF, but this doesn't work always.

**Reply by anthrotype** [reply]
I rebased that qu2cu PR branch onto the latest cu2qu main branch, you can find it at https://github.com/googlefonts/cu2qu/compare/qu2cu-rebased

there's two additional cu2qu.pens called respectively Qu2CuPen (segment pen, works with glyph.draw(pen) method ) and Qu2CuPointPen (point pen, works with glyph.drawPoints(pen)). Try see if they work for you

**Reply by ftCLI** [reply]
This is my actual code (a subclass of TTFont):

```python
    def get_cff_charstrings(self) -> dict:
        charstrings = {}
        glyph_set = self.getGlyphSet()

        for k, v in glyph_set.items():

            # Remove overlaps and fix contours direction
            pathops_path = pathops.Path()
            pathops_pen = pathops_path.getPen(glyphSet=glyph_set)
            try:
                glyph_set[k].draw(pathops_pen)
                pathops_path.simplify()
            except TypeError:
                pass

            # Draw the glyph with T2CharStringPen and get the charstring
            t2_pen = T2CharStringPen(v.width, glyphSet=glyph_set)
            pathops_path.draw(t2_pen)
            charstring = t2_pen.getCharString()
            charstrings[k] = charstring

        return charstrings
```

Where can I use the Qu2CuPen?

**Reply by anthrotype** [reply]
it's a filter pen that wraps another pen, ie. takes another pen as parameter, so it would go here:

```python
t2_pen = T2CharStringPen(v.width, glyphSet=glyph_set)
qu2cu_pen = Qu2CuPen(t2_pen)
pathops_path.draw(qu2cu_pen)
```

**Reply by ftCLI** [reply]
I made a test with IBM Plex Serfi. Unfortunately, the outer contours disappear from all the 'o' characters.

![Cattura](https://user-images.githubusercontent.com/83063506/206741015-114fc003-f9e7-4ca5-8a74-a78e4a93554f.PNG)

This is what happens to the lighter styles:
![image](https://user-images.githubusercontent.com/83063506/206742237-be41c4df-0d89-4935-bffa-7b6c1be49dc5.png)

Nice job anyway, it's really fast. I hope that you'll find the time to finish it one day.

Here are the converted files: [otf.zip](https://github.com/fonttools/fonttools/files/10196211/otf.zip)

**Reply by anthrotype** [reply]
hm yeah.. it only works somewhat reliably with quadratic curves that have in turn been generated by cu2qu basically, it currently doesn't have a way to approximate within given tolerance but simply assumes that you want exactly one cubic bezier for each TrueType quadratic spline in the input, sometimes you may need more than 1..

**Reply by anthrotype** [reply]
though I suspect the outer contours disappearing from the 'o' might be the result of skia-pathops simplify method, rather than qu2qu.

**Reply by ftCLI** [reply]
Same result even if I comment the `simplify()` line.

Anyway, here's my script (that is not the old one in my repository) in case you or someone else wants to try: [ftCLI.zip](https://github.com/fonttools/fonttools/files/10196574/ftCLI.zip)

The specific command is

`ftcli converter ttf2otf PATH`, where PATH can be a file or a folder:

**Reply by ftCLI** [reply]
My bad. The pen works fine (except for the composite glyphs, that still have a lot of nodes) without pathops. Once the file is saved, reopening it and fixing the contours direction with pathops works.

**Comment by behdad** [comment]
I'm wondering what the subject line has to do with the actual topic. :)

**Reply by ftCLI** [reply]
Changed
